### DISCUSSION TITLE
Interpret the “attractiveness of a pattern” in an image and assign a numeric value 1-100 for image score

### DISCUSSION THREAD

**Discussion by jarj1** [ImageMagick/ImageMagick#2127] [Help]
IM Windows/CLI Version: ImageMagick 6.9.10-11 Q16 x86_64
And/or            PHP 3.4.4  ImageMagick 7.0.7-11 Q16 x64 2017-11-23 

Problem: An image pattern attractiveness is defined as having larger “blocks” of same or similar colour and that those blocks not being dull or a reflection.  The closer to the center the block is the higher the weighting.  Dull blocks or grainy blocks detract from attractiveness.

Attempts so far have been a mixture of php and cli
PHP: loop through each pixel and if it is one of significance (S>30 and not reflection or background) floodfilling from that pixel and counting the effected pixels and adding the block of colour to an array and then sorting the array and determining a result.  This has worked to a point however the image has to be very simplified first for the program to finish within an hour and the simplification looses details and so areas which are grainy and similar erroneously get a high result as the program sees them as a single colour area (see PHP_01_sim.png) in the simplified image.

CLI:Ideas were to
1) Remove reflection (convert input.png -fuzz 12% -transparent white no_refx.png)
2) Do something similar as No1 to remove low Saturation pixels and pixels with a low Value
3) Either:

A.  Run canny over the output and count the white pixels vs black as the white would edge the blocks of colour – more white = finer/grainier pattern [note: somehow the image would need to get penalty points for blocks of low saturation or low value – background and reflection exempt] Maybe in No2 the low saturation/value blocks could be replaced with a fine checker-board pattern which would get automatically picked up by the canny and lower the attractiveness return value.

B. Using values from identify -verbose manipulated.png | head -61 | tail -7 or other image statistics?

Images desired result something like below: 
  No1 20-30%
  No2 50-60%
  No3 80-90%
  No4 20-30%
  No5 70-80%
  No6 30-40%

Code or just ideas much appreciated
TIA

![CLI_01_HQ](https://user-images.githubusercontent.com/66709392/84226381-f2dd8e80-ab24-11ea-8f95-598c3116911b.png)
![CLI_02_HQ](https://user-images.githubusercontent.com/66709392/84226386-f4a75200-ab24-11ea-8f0a-1a79fe4008e5.png)
![CLI_03_HQ](https://user-images.githubusercontent.com/66709392/84226390-f7a24280-ab24-11ea-8232-f893e219a66e.png)
![CLI_04_HQ](https://user-images.githubusercontent.com/66709392/84226392-f8d36f80-ab24-11ea-8a38-88ffc78cad41.png)
![CLI_05_HQ](https://user-images.githubusercontent.com/66709392/84226397-fb35c980-ab24-11ea-8d98-3f7237922896.png)
![CLI_06_HQ](https://user-images.githubusercontent.com/66709392/84226399-fc66f680-ab24-11ea-8f46-c99ce48cf4c6.png)
![PHP_01](https://user-images.githubusercontent.com/66709392/84226404-fec95080-ab24-11ea-942d-15cfe357a6ea.png)
![PHP_01_sim](https://user-images.githubusercontent.com/66709392/84226406-ff61e700-ab24-11ea-8d1e-6a707bed3550.png)
![PHP_02](https://user-images.githubusercontent.com/66709392/84226407-ff61e700-ab24-11ea-8060-70892f58f771.png)
![PHP_02_sim](https://user-images.githubusercontent.com/66709392/84226408-fffa7d80-ab24-11ea-9ec2-b9eeaab69001.png)
![PHP_03](https://user-images.githubusercontent.com/66709392/84226409-00931400-ab25-11ea-8281-6c1fc415120b.png)
![PHP_03_sim](https://user-images.githubusercontent.com/66709392/84226410-00931400-ab25-11ea-86ca-110fc938e2ee.png)
![PHP_04](https://user-images.githubusercontent.com/66709392/84226411-012baa80-ab25-11ea-94f6-c5c274c61909.png)
![PHP_04_sim](https://user-images.githubusercontent.com/66709392/84226415-01c44100-ab25-11ea-8ec4-d60748e18e02.png)

**Comment by snibgo** [comment]
This problem looks suitable for machine learning, if you can train it with enough image/score pairs.

>> ... defined as having larger “blocks” of same or similar colour ...

What defines a "block"? How do we know where the edges of a block are?

>> ... those blocks not being dull or a reflection.

What defines "dull" and "reflection"?

On PHP performance: perhaps you are looping through all the pixels in PHP. This will be slow. The trick is either to use a compiled language, or to leverage the built-in operations of IM.

These look like photos of polished pebbles.

We seem to be aiming for light pixels with high saturation. Suppose a pebble was entirely pure yellow. Would that score 100%? If not, why not? Similarly, a black pebble would be 0%?

Perhaps we can calculate a score for each pixel. Then calculate the overall score as the average of each pixel, weighted by the distance of the pixel from the centre. (Centre = centroid?)

So, how do we score each pixel? It is a function of lightness and saturation (eg the C and L channels of HCL), but also similarity to nearby pixels. If this pixel is different to nearby pixels, this lowers its score. This suggests a sliding window, so each pixel is compared to the average window  ("local average") in a sliding rectangle that is 21x21 pixels, or whatever. This is crude but fast, and perhaps good enough for initial trials. More sophisticated: we compare to nearby pixels in a circular window, with gaussian distribution, which be can be made with a blur.

We might define "difference between this pixel and the local average" is in CIE Lab space, and the score is sqrt(dL^2 + da^2 + db^2).

Do these thoughts cover your requirements?

**Reply by jarj1** [reply]
Thanks for your help snibgo much appreciated.
I know little about machine learning but I imagine it would be a lot of coding and 1,000 or 100 of thousands images?

blocks being defined as the same colour with a bit of fuzz but uninterrupted and so why No3 is much higher than No1.  If I run  "convert in.png  -canny 0x1+10^%+10^% out.png" on the two images I get the below:
![i_00000c](https://user-images.githubusercontent.com/66709392/84261396-09a0d700-ab5f-11ea-9ab4-3802ac89d2ec.png)
![i_00000c](https://user-images.githubusercontent.com/66709392/84261421-145b6c00-ab5f-11ea-9b36-cc6d4200d557.png)
showing edges of blocks/how busy or simple the pattern is.

The actual inputs for determining what is reflection and what is dull would probably have to be determined for each image based on some average like the %k.  I'm happy I can remove the reflection for example on the No1 image if I run "convert in.png -fuzz 12% -fill black -opaque white   out.png" which identifies and removes the two reflections top and bottom.
![i_00000r](https://user-images.githubusercontent.com/66709392/84262707-69987d00-ab61-11ea-870b-3f9c4e474472.png)

I would like a way to do something similar but instead of finding white I want to select all pixels where the Saturation is below 18% and then repeat filtering off where the Value is below 75%

Yes if a pebble was entirely yellow, red, green or blue it would score 100% if it were half blue and half red and those halves being for argument sake the top and the bottom it would score 99.9999% however if it were half red and half blue comprising of a blue background with lots of tiny red dots it would score 10% or less.  Yes a black pebble would be 0%

I'm interested in the each pixel 21x21 window crude and fast sound good to me but as for actually coding something to achieve it - way beyond my IM skills, can it be done on the CLI or would it have to be done in PHP?

I'm not super fussed with the center weighting if its going to really complicate things - It is a nice to have but not essential.  If the worst came to the worst I could cut the images up into a center core, middle mantle and outer crust - score them individually and avg the results weighting the core, then mantle and least the crust - clumsy and childish I know but it would suffice

**Comment by snibgo** [comment]
Centre weighting is easy. We create a gradient image, white at the pebble's centre, and black around the edge, centGrad.png below. This is then a multiplier to get weighting.

The reflections of the lighting have low saturation, so C in HCL is low.

I create two other images: cl.png is the geometric mean of the C and L channels from HCL. labstd.png is the standard deviation of a sliding 11x11 window, tweaked, so the result is dark where the SD is high (there is lots of variation, eg grainy) and the result is light where the SD is low (there is little variation).

I take the geometric mean of these three images. The average of the opaque pixels gives the numeric score.

The scores, on a scale of 0.0 to 1.0, are:

stone1.png 0.305936
stone2.png 0.317948
stone3.png 0.32414
stone4.png 0.275185
stone5.png 0.326447
stone6.png 0.323325

The ranking of these is similar to the ranking in your desired results.

My Windows BAT script, stone2.bat, is:
```
echo off

set INFILE=%1

%IMG7%magick ^
  %INFILE% ^
  -alpha extract ^
  alph.png

rem The multiplier and power are arbitrary.
%IMG7%magick ^
  %INFILE% ^
  -alpha off ^
  -colorspace Lab -set colorspace sRGB ^
  -statistic standard_deviation 11x11 ^
  -grayscale RMS ^
  -evaluate Multiply 2 -evaluate Pow 0.4 ^
  -negate ^
  labstd.png

%IMG7%magick ^
  %INFILE% ^
  -colorspace HCL ^
  -set colorspace sRGB ^
  -channel GB -separate +channel ^
  -compose Multiply -composite ^
  -evaluate Pow 0.5 ^
  cl.png

%IMG7%magick ^
  alph.png ^
  -morphology Distance Chebyshev:"1,1" ^
  -auto-level ^
  centGrad.png

for /F "usebackq" %%L in (`%IMG7%magick ^
  labstd.png ^
  cl.png ^
  centGrad.png ^
  -background White ^
  -compose Multiply -layers Flatten ^
  -evaluate Pow %%[fx:1/3] ^
  +write mult.png ^
  -scale "1x1^!" ^
  -format "score=%%[fx:mean]"\n ^
  info:`) do set %%L

echo %INFILE% %score%

echo on
```
When the script does what you want, the individual commands can be combined.

**Comment by jarj1** [comment]
I've made some minor batch type changes so I can get the output images into one image.  I got the latest IM7 version and ran the script, I am getting different results?

I haven't changed the multiplier or power - I want to get the same results as you first before experimenting:

\web\im\magick -version
Version: ImageMagick 7.0.10-18 Q16 x64 2020-06-09 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 180040629
Features: Cipher DPC HDRI OpenCL
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib


FOR /L %i IN (1, 1, 6 ) DO get_pattern %i
Stone 1 0.652964
Stone 2 0.65897
Stone 3 0.662066
Stone 4 0.637588
Stone 5 0.663219
Stone 6 0.661658

![output](https://user-images.githubusercontent.com/66709392/84330280-51107d00-abca-11ea-8e00-9dcdb7dedb64.png)

`@echo off
set IMG7=C:\web\im\
set STONE=%1

set INFILE=%STONE%i_00001.png

%IMG7%magick ^
  %INFILE% ^
  -alpha extract ^
  %STONE%alph.png

rem The multiplier and power are arbitrary.
%IMG7%magick ^
  %INFILE% ^
  -alpha off ^
  -colorspace Lab -set colorspace sRGB ^
  -statistic standard_deviation 11x11 ^
  -grayscale RMS ^
  -evaluate Multiply 2 -evaluate Pow 0.4 ^
  -negate ^
  %STONE%labstd.png

%IMG7%magick ^
  %INFILE% ^
  -colorspace HCL ^
  -set colorspace sRGB ^
  -channel GB -separate +channel ^
  -compose Multiply -composite ^
  -evaluate Pow 0.5 ^
  %STONE%cl.png

%IMG7%magick ^
  %STONE%alph.png ^
  -morphology Distance Chebyshev:"1,1" ^
  -auto-level ^
  %STONE%centGrad.png

for /F "usebackq" %%L in (`%IMG7%magick ^
  %STONE%labstd.png ^
  %STONE%cl.png ^
  %STONE%centGrad.png ^
  -background White ^
  -compose Multiply -layers Flatten ^
  -evaluate Pow %%[fx:1/3] ^
  +write %STONE%mult.png ^
  -scale "1x1^!" ^
  -format "score=%%[fx:mean]"\n ^
  info:`) do set %%L

echo Stone %STONE% %score%`

**Comment by snibgo** [comment]
You might need "setlocal enabledelayedexpansion" at the top.

**Reply by jarj1** [reply]
I getting the same higher numbers  with enabledelayedexpansion as with without

Do your CL labstd and mult images look similar to the ones in the collage above?  Could it be my IM7 version?

`setlocal enabledelayedexpansion
@echo off

set IMG7=C:\web\im\
set STONE=%1

set INFILE=%STONE%i_00001.png
...`

**Comment by jarj1** [comment]
I've been playing around with the script, but have hit a bit of a wall...  Is there any way to make that "window" more dynamic so that the larger the area of alike pixels the brighter its center (score higher)?
